src/time-picker/index.tsx

```diff
@@ -335,7 +335,7 @@ export const TimePicker = factory(function TimePicker({
 				}
 			}
 
-			isValid = validationMessages.length === 0;
+			isValid = isValid === undefined ? validationMessages.length === 0 : isValid;
```

Comment:
This was the root of the issue: The underlying `TextInput` updates `inputIsValid`, but then local validation and `validationMessages.length` ended up overwriting this before triggering `onValidate`.